src/backend/distributed/utils/resource_lock.c

```diff
@@ -122,6 +123,93 @@ lock_shard_resources(PG_FUNCTION_ARGS)
 }
 
 
+/*
+ * LockShardListResourcesOnWorkers acquires the resource locks for the specified shards on
+ * metadata workers. Note that the function does not sort the shard list, therefore the
+ * caller should sort the shard list in order to avoid deadlocks.
+ */
+void
+LockShardListResourcesOnWorkers(TargetWorkerSet targetWorkerSet, LOCKMODE lockmode,
+								List *shardIntervalList)
+{
+	StringInfo lockCommand = makeStringInfo();
+	ListCell *shardIntervalCell = NULL;
+	List *workerNodeList = ActivePrimaryNodeList();
+	int processedShardPlacementCount = 0;
+	int totalShardPlacementCount = list_length(shardIntervalList);
+
+	/* In order to get lock on the first metadata node, we need to sort them */
+	if (targetWorkerSet == FIRST_WORKER_WITH_METADATA)
+	{
+		WorkerNode *firstWorkerNode = NULL;
+		workerNodeList = SortList(workerNodeList, CompareWorkerNodesByGroupId);
+
+		firstWorkerNode = (WorkerNode *) linitial(workerNodeList);
+		if (firstWorkerNode->groupId == GetLocalGroupId())
+		{
+			ereport(WARNING, (errmsg(
+								  "Since the query is sent from first worker, won't open connection to itself.")));
```

Comment:
Otherwise we can face with self deadlock. Both the transaction will try to get the lock and the distributed transaction opened by that node on itself will try to get that lock.